src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs

```diff
@@ -0,0 +1,313 @@
+                    },
+                },
+            };
+
+            var hasJsonBody = false;
+
+            foreach (var parameter in methodInfo.GetParameters())
+            {
+                var parameterDescription = CreateApiParameterDescription(parameter, routeEndpoint.RoutePattern);
+
+                if (parameterDescription.Source == BindingSource.Body)
+                {
+                    hasJsonBody = true;
+                }
+
+                apiDescription.ParameterDescriptions.Add(parameterDescription);
+            }
+
+            AddSupportedRequestFormats(apiDescription.SupportedRequestFormats, hasJsonBody, routeEndpoint.Metadata);
+            AddSupportedResponseTypes(apiDescription.SupportedResponseTypes, methodInfo.ReturnType, routeEndpoint.Metadata);
+
+            return apiDescription;
+        }
+
+        private static ApiParameterDescription CreateApiParameterDescription(ParameterInfo parameter, RoutePattern pattern)
+        {
+            var (source, name) = GetBindingSourceAndName(parameter, pattern);
+
+            return new ApiParameterDescription
+            {
+                Name = name,
+                ModelMetadata = CreateModelMetadata(parameter.ParameterType),
+                Source = source,
+                DefaultValue = parameter.DefaultValue,
+            };
+        }
+
+        // TODO: Share more of this logic with RequestDelegateFactory.CreateArgument(...) using RequestDelegateFactoryUtilities
+        // which is shared source.
+        private static (BindingSource, string) GetBindingSourceAndName(ParameterInfo parameter, RoutePattern pattern)
+        {
+            var attributes = parameter.GetCustomAttributes();
+
+            if (attributes.OfType<IFromRouteMetadata>().FirstOrDefault() is { } routeAttribute)
+            {
+                return (BindingSource.Path, routeAttribute.Name ?? parameter.Name ?? string.Empty);
+            }
+            else if (attributes.OfType<IFromQueryMetadata>().FirstOrDefault() is { } queryAttribute)
+            {
+                return (BindingSource.Query, queryAttribute.Name ?? parameter.Name ?? string.Empty);
+            }
+            else if (attributes.OfType<IFromHeaderMetadata>().FirstOrDefault() is { } headerAttribute)
+            {
+                return (BindingSource.Header, headerAttribute.Name ?? parameter.Name ?? string.Empty);
+            }
+            else if (parameter.CustomAttributes.Any(a => typeof(IFromBodyMetadata).IsAssignableFrom(a.AttributeType)))
+            {
+                return (BindingSource.Body, parameter.Name ?? string.Empty);
+            }
+            else if (parameter.CustomAttributes.Any(a => typeof(IFromServiceMetadata).IsAssignableFrom(a.AttributeType)) ||
+                     parameter.ParameterType == typeof(HttpContext) ||
+                     parameter.ParameterType == typeof(CancellationToken) ||
+                     parameter.ParameterType.IsInterface)
```

Comment:
TODO: rebase and copy the new `IServiceProviderIsService` logic used in RequestDelegateFactory. Unfortunately this breaks my out-of-repo sample, since we don't have a full SDK with this changes available from dotnet/installers yet 😢